File: day-7/part-1.py
FIRST = len('Step ')
SECOND = len('Step C must be finished before step ')

class Node:
 def __init__(self, key):
   self.incoming = []
   self.outgoing = []
   self.key = key

 def __repr__(self):
   return self.key

 def __str__(self):
   return self.key

def getOrAdd(graph, key):
  if key not in graph:
    graph[key] = Node(key)
  return graph[key]

with open('input') as f:
  lines = f.readlines()
  pairs = map(lambda line: (line[FIRST], line[SECOND]), lines)
  graph = {}
  for (a, b) in pairs:
    nodeA = getOrAdd(graph, a)
    nodeB = getOrAdd(graph, b)
    nodeA.outgoing.append(b)
    nodeB.incoming.append(a)
  
  queue = filter(lambda node: len(node.incoming) == 0, graph.values())
  for node in queue:
    del graph[node.key]
  
  done = []
  while len(queue):
    queue.sort(key=lambda node: node.key)
    current = queue.pop(0)
    done.append(current.key)
    for out in current.outgoing:
      if out not in graph:
        continue
      other = graph[out]
      other.incoming.remove(current.key)
      if len(other.incoming) == 0:
        queue.append(other)
        del graph[out]
  print "".join(done)

File: day-07/part-1.js
const { readLines } = require('../util.js');

const LINE = 'Step X must be finished before step ';

function getNode(map, id) {
  let node = map.get(id);
  if (!node) {
    map.set(id, node = {id, requirements:[], nodes: []});
  }
  return node;
}
const graph = readLines().map(line => {
      let requirement = line.slice('Step '.length)[0];
      let node = line.slice(LINE.length)[0];
      return {requirement, node};
    }).reduce((graph, edge) => {
      let r = getNode(graph, edge.requirement);
      let n = getNode(graph, edge.node);
      n.requirements.push(r.id);
      r.nodes.push(n.id);
      return graph;
    }, new Map());

let queue = [...graph.entries()]
              .filter(([id, node]) => !node.requirements.length)
              .map(([id, node]) => node);
queue.forEach(node => graph.delete(node.id));
let order = [];
while (queue.length) {
  queue.sort((a, b) => a.id < b.id ? -1 : 1);
  let node = queue.shift();
  order.push(node.id);
  node.nodes.forEach(other => {
    let onode = graph.get(other);
    onode.requirements.splice(onode.requirements.indexOf(node.id), 1);
    if (onode.requirements.length === 0) {
      graph.delete(other);
      queue.push(onode);
    }
  });
}
console.log(order.join(''));

